### ISSUE TITLE
/venv directory on repo containers not syncing before openstack-setup play runs on upgrades

### ISSUE THREAD

**Issue by jacobwagner** [rcbops/rpc-openstack#1007]
##### What version (commit) do you encounter this issue on?
12.1.0rc3
##### What type of issue is this (bug, feature, documentation, regression)?
bug
##### What is the expected behavior?
all 3 repo servers should contain all proper directories under /var/www/repo
##### What is actually happening?
the 2 slave repos are not getting full syncs of the /venv repo
##### How can we reproduce this issue?
install kilo lab (11.1.6), run rpc-openstack upgrade script (we use haproxy), should fail on os-keystone install
##### What is your evaluation on the importance of this issue, and why?
prio-1, as upgrades that use the /venv repo folder will fail
###### Remove branches (lines) you do not wish to target.
- [x] Master
- [x] liberty-12.2
- [x] liberty-12.1
```
root@node89_repo_container-58161321:/var/www/repo# ls -al
total 272
drwxr-sr-x   9 nginx www-data   4096 Apr 18 22:09 .
drwxr-sr-x   5 nginx www-data   4096 Apr 18 20:53 ..
-rw-r--r--   1 nginx www-data    486 Apr 18 22:09 index.html
drwxr-sr-x   2 nginx www-data 208896 Apr 18 02:43 links
drwxr-sr-x  51 nginx www-data   4096 Apr 17 15:01 openstackgit
drwxr-sr-x  36 nginx www-data   4096 Apr 18 22:09 os-releases
drwxr-sr-x 671 nginx www-data  32768 Apr 18 22:09 pools
drwxr-sr-x   3 nginx www-data   4096 Apr 18 22:09 python_packages
drwxr-sr-x   2 nginx www-data   4096 Apr 18 22:09 reports
drwxr-sr-x   3 nginx www-data   4096 Apr 18 22:09 rpc-releases
```

```
root@node98_repo_container-3bd9ec93:/var/www/repo# ls -al
total 280
drwxr-sr-x  10 nginx www-data   4096 Apr 19 01:44 .
drwxr-sr-x   5 nginx www-data   4096 Apr 18 22:09 ..
-rw-r--r--   1 nginx www-data    486 Apr 18 22:09 index.html
drwxr-sr-x   2 nginx www-data 208896 Apr 19 01:44 links
drwxr-sr-x  52 nginx www-data   4096 Apr 19 01:28 openstackgit
drwxr-sr-x  36 nginx www-data   4096 Apr 19 01:29 os-releases
drwxr-sr-x 673 nginx www-data  36864 Apr 19 01:41 pools
drwxr-sr-x   3 nginx www-data   4096 Apr 18 22:09 python_packages
drwxr-sr-x   2 nginx www-data   4096 Apr 18 22:09 reports
drwxr-sr-x   3 nginx www-data   4096 Apr 18 22:09 rpc-releases
drwxr-sr-x   3 nginx www-data   4096 Apr 19 01:44 venvs
```

```
root@node84_repo_container-36dc97ef:/var/www/repo# ls -al
total 304
drwxr-sr-x   9 nginx www-data   4096 Apr 18 22:09 .
drwxr-sr-x   5 nginx www-data   4096 Apr 18 20:53 ..
-rw-r--r--   1 nginx www-data    486 Apr 18 22:09 index.html
drwxr-sr-x   2 nginx www-data 237568 Apr 18 02:43 links
drwxr-sr-x  51 nginx www-data   4096 Apr 17 15:01 openstackgit
drwxr-sr-x  36 nginx www-data   4096 Apr 18 22:09 os-releases
drwxr-sr-x 671 nginx www-data  36864 Apr 18 22:09 pools
drwxr-sr-x   3 nginx www-data   4096 Apr 18 22:09 python_packages
drwxr-sr-x   2 nginx www-data   4096 Apr 18 22:09 reports
drwxr-sr-x   3 nginx www-data   4096 Apr 18 22:09 rpc-releases
```

These missing repos are causing the LB to return 404 when it accesses the 2 non syncd repos

```
root@node84:~# curl http://172.29.236.121:8181/venvs/12.0.10/ubuntu/keystone-12.0.10.checksum
<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.4.6 (Ubuntu)</center>
</body>
</html>
root@node84:~# curl http://172.29.236.121:8181/venvs/12.0.10/ubuntu/keystone-12.0.10.checksum
31227e86248ed75238c32a32b3bbdfc9d43e7d0d
root@node84:~# curl http://172.29.236.121:8181/venvs/12.0.10/ubuntu/keystone-12.0.10.checksum
<html>
<head><title>404 Not Found</title></head>
<body bgcolor="white">
<center><h1>404 Not Found</h1></center>
<hr><center>nginx/1.4.6 (Ubuntu)</center>
</body>
</html>
```

**Comment by jacobwagner** [comment]
UPDATE

```
root@node98_repo_container-3bd9ec93:/var/log/lsyncd# tail -n 10 lsyncd.log
Tue Apr 19 01:23:27 2016 Normal: Startup of "/var/www/repo/" finished.
Tue Apr 19 01:24:56 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.237.91:/var/www/repo/
Tue Apr 19 01:24:56 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.237.21:/var/www/repo/
ssh: connect to host 172.29.237.21 port 22: No route to host
ssh: connect to host 172.29.237.91 port 22: No route to host
rsync: connection unexpectedly closed (0 bytes received so far) [sender]
rsync: connection unexpectedly closed (0 bytes received so far) [sender]
rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.0]
rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.0]
Tue Apr 19 01:24:59 2016 Error: Temporary or permanent failure on startup of "/var/www/repo/". Terminating since "insist" is not set.
```

looks like it couldnt reach the other hosts to sync

**Comment by jacobwagner** [comment]
UPDATE

a restart of lysncd on the main repo server worked

all 3 repo servers are syncd

**Comment by git-harry** [comment]
I have targeted at rc4.

I am not aware of any previous reports of this issue. The workaround (restarting lsyncd) is fairly simple. Based on these points it may be better to flag this as a known issue and fix in a future release. If that's the case we can remove it from the rc4 milestone.

**Comment by git-harry** [comment]
We need a recommendation from whoever works this issue to determine the best course of action. I favour making this a known issue unless there good justification to do otherwise.

**Comment by jacobwagner** [comment]
So i cant get any good timestamp info on the syslogs, but what i can get makes it looks like the main repo container is restarting before the 2 slaves, lysncd is starting and cant connect to the other 2 so the serivce stops, so when the venvs dir gets created it doesnt get synced.

This is only happening on upgrades due to the repo server existing and lsyncd running on kilo. We didnt see it in greenfields.

Maybe ensure that lysncd is running before attempting or assuming the sync is happening?

**Comment by b3rnard0** [comment]
@d34dh0r53 and @prometheanfire Can you all please investigate this issue to determine what's going and what's needed. This is looking like a known issue at this point.

**Comment by d34dh0r53** [comment]
@b3rnard0 yeah, been looking at this today, forgot to update the issue.

**Comment by hughsaunders** [comment]
I've done some testing and lsyncd quits on startup if it cant connect to all the slaves. There is however an option "-insist" to make lsyncd retry until it can connect to them all. I'll raise an OSA bug and get that added.

**Comment by git-harry** [comment]
@d34dh0r53 what have you found from your investigations?

**Comment by hughsaunders** [comment]
https://bugs.launchpad.net/openstack-ansible/+bug/1572433

**Comment by hughsaunders** [comment]
@git-harry release note blurb: 

> When multiple repo servers are deployed they are kept in sync via lsyncd, on startup lsyncd on the primary repo server (repo_all[0]) performs a recursive rsync to all slaves, after that only changed files are synced. If a slave is not available when lsyncd performs the initial rsync, lsyncd will not retry and will exit, potentially leaving the repo containers out of sync. This may be encountered when restarting servers/containers or upgrading (which includes restarts). If this issue is encountered, the workaround is to restart lsyncd on the primary repo server once the others are all available. To detect this issue look for the following message in the lsyncd log on the primary repo server: 'Error: Temporary or permanent failure on startup of "/var/www/repo/". Terminating since "insist" is not set.'

**Comment by git-harry** [comment]
thanks @hughsaunders

**Comment by d34dh0r53** [comment]
I had some AIO issues and didn't get as far as @hughsaunders, since this is in progress I'll keep an eye on the upstream bug and provide reviews to get it merged ASAP.

**Comment by hughsaunders** [comment]
Patch has now merged in OSA liberty, so should be available to RPC in the next OSA SHA bump.

**Comment by jacobwagner** [comment]
saw this again in the iad lab upgrade

```
root@569038-infra03_repo_container-e693aded:~# tail -n 100 /var/log/lsyncd/lsyncd.log
Thu Apr 21 15:26:57 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.239.66:/var/www/repo/
Connection to 172.29.238.247 closed by remote host.
rsync: connection unexpectedly closed (477 bytes received so far) [sender]
rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.0]
Thu Apr 21 15:26:58 2016 Error: Temporary or permanent failure on startup of "/var/www/repo/". Terminating since "insist" is not set.
Thu Apr 21 15:29:24 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.238.247:/var/www/repo/
Thu Apr 21 15:29:24 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.239.66:/var/www/repo/
ssh: connect to host 172.29.239.66 port 22: No route to host
ssh: connect to host 172.29.238.247 port 22: No route to host
rsync: connection unexpectedly closed (0 bytes received so far) [sender]rsync: connection unexpectedly closed (0 bytes received so far) [sender]

rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.0]rsync error: unexplained error (code 255) at io.c(226) [sender=3.1.0]

Thu Apr 21 15:29:28 2016 Error: Temporary or permanent failure on startup of "/var/www/repo/". Terminating since "insist" is not set.
Thu Apr 21 15:49:53 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.238.247:/var/www/repo/
Thu Apr 21 15:49:53 2016 Normal: recursive startup rsync: /var/www/repo/ -> 172.29.239.66:/var/www/repo/
Thu Apr 21 15:50:03 2016 Normal: Startup of "/var/www/repo/" finished.
Thu Apr 21 15:50:03 2016 Normal: Startup of "/var/www/repo/" finished.
Thu Apr 21 15:50:11 2016 Normal: Calling rsync with filter-list of new/modified files/dirs
```

good news is as long as you bump it after the container restart you should be ok...i did it live during the upgrade

**Comment by d34dh0r53** [comment]
Upstream has merged, the next SHA bump should include this fix upstream.

**Comment by b3rnard0** [comment]
Blocked until the SHA bump.

**Comment by prometheanfire** [comment]
lolnope, fixed in #1054 #1055 #1056

**Comment by prometheanfire** [comment]
bumped
